### PR TITLE
Add config to store objects on stack vs. scratch buffer

### PR DESCRIPTION
**Pull request by ryantimwilson**
This PR adds a configurable limit `on_stack_limit` - default = 32 - to allow the user to specify whether tuples/fmt strings/str calls (and soon map keys/values) should be stored on the stack or in a scratch buffer. This allows us to balance supporting large strings/objects while avoiding making the ResourceAnalyser code too complex by trying to accurately estimating memory for map values/keys: https://github.com/bpftrace/bpftrace/pull/3483

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -138,6 +138,7 @@ void usage()
   std::cerr << "    BPFTRACE_MAX_TYPE_RES_ITERATIONS  [default: 0] number of levels of nested field accesses for tracepoint args" << std::endl;
   std::cerr << "    BPFTRACE_PERF_RB_PAGES            [default: 64] pages per CPU to allocate for ring buffer" << std::endl;
   std::cerr << "    BPFTRACE_STACK_MODE               [default: bpftrace] Output format for ustack and kstack builtins" << std::endl;
+  std::cerr << "    BPFTRACE_STACK_THRESHOLD          [default: 32] max bytes to store objects on BPF stack vs. global variable" << std::endl;
```
Comment (jordalgo) [inline src/main.cpp:141]:
```suggestion
  std::cerr << "    BPFTRACE_STACK_THRESHOLD          [default: 32] max bytes of individual objects to store them on the BPF stack vs. a global variable" << std::endl;
```

File: man/adoc/bpftrace.adoc [hunk 2]
```diff
@@ -3571,6 +3571,15 @@ Available modes/formats:
 
 This can be overwritten at the call site.
 
+=== stack_threshold
+
+Default: 32
+
+The maximum length (in bytes) where objects will be stored on the BPF stack instead of a scratch global variable.
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:3578]:
```suggestion
The maximum size (in bytes) of individual objects that will be stored on the BPF stack. If they are larger than this threshold they will be stored in a scratch global variable.
```

File: tests/codegen/scratch_buffer.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#include "common.h"
+
+namespace bpftrace {
+namespace test {
+namespace codegen {
+
+TEST(codegen, big_tuple_scratch_buf)
+{
+  test("kprobe:f { (1, \"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\") }",
```
Comment (jordalgo) [inline tests/codegen/scratch_buffer.cpp:9]: Maybe these tests should explicitly set the config variable so if we change the default they don't change. And maybe have two codegens (under and over threshold) for each fmt and tuples?

**Review by jordalgo — COMMENTED** [review]
Few small things but overall looks good.

File: src/main.cpp (hunk 1 above)
Comment (danobi) [reply src/main.cpp:141]: "global variable" kinda leaks an unnecessary implementation detail. Maybe say "preallocated memory"?

Comment (danobi) [reply src/main.cpp:141]: nit: "stack threshold" seems backwards (implying a minimum for the stack). maybe "offstack threshold"?

File: src/ast/passes/codegen_llvm.h [hunk 4]
```diff
@@ -248,6 +249,13 @@ class CodegenLLVM : public Visitor {
 
   Value *createFmtString(int print_id);
 
+  Value *createTupleAllocation(const SizedType &tuple_type,
+                               const std::string &name,
+                               const location &loc);
+  Value *createGetFmtStringArgsAllocation(StructType *struct_type,
+                                          const std::string &name,
+                                          const location &loc);
+
```
Comment (danobi) [inline src/ast/passes/codegen_llvm.h:258]: Wondering if there's a way to prevent proliferation of wrappers like this. Seems like it would be good to keep the "stack or offstack" decision in a central place so new offstack types can't easily forget to check on stack_threshold.

I can't think of any better ideas at the moment so just consider this a non-blocking idea to ponder.

**Review by danobi — COMMENTED** [review]
overall lgtm.

File: tests/codegen/scratch_buffer.cpp (hunk 3 above)
Comment (danobi) [reply tests/codegen/scratch_buffer.cpp:9]: +1, testing both paths is good idea

File: src/main.cpp (hunk 1 above)
Comment (jordalgo) [reply src/main.cpp:141]: I think maybe both are ambiguous. How about "on_stack_limit" or "on_stack_max" or "off_stack_min"

File: src/ast/passes/codegen_llvm.h (hunk 4 above)
Comment (ryantimwilson) [reply src/ast/passes/codegen_llvm.h:258]: Yea the problem is we depend on iterating `async_ids` if using scratch, else not iterating it. So its hard to move into irbuilderbpf.cpp. I guess we could do this with a callback...let me experiment with that approach.

File: src/main.cpp (hunk 1 above)
Comment (ryantimwilson) [reply src/main.cpp:141]: Lets discuss naming offline

Comment (ryantimwilson) [reply src/main.cpp:141]: on_stack_limit seems to be the choice, lets go with that

**Comment by ryantimwilson** [comment]
@danobi @jordalgo 

Updates:
1. Changed name to `off_stack_limit` and updated relevant descriptions
2. Updated strings to use stack if max_strlen > on_stack_limit
3. Added codegen tests for stack vs. scratch buffer
4. Refactored to use common stack vs. scratch buffer function in irbuilderbpf

File: src/ast/passes/codegen_llvm.h (hunk 4 above)
Comment (ryantimwilson) [reply src/ast/passes/codegen_llvm.h:259]: Addressed in latest version using callbacks

File: tests/codegen/scratch_buffer.cpp (hunk 3 above)
Comment (ryantimwilson) [reply tests/codegen/scratch_buffer.cpp]: Done in latest!

File: man/adoc/bpftrace.adoc [hunk 5]
```diff
@@ -3556,6 +3556,16 @@ The possible options are:
 - `warn` - print a warning but continue execution
 - `ignore` - silently ignore missing probes
 
+=== on_stack_limit
+
+Default: 32
+
+The maximum size (in bytes) of individual objects that will be stored on the BPF stack. If they are larger than this threshold they will be stored in
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:3563]: nit
```suggestion
The maximum size (in bytes) of individual objects that will be stored on the BPF stack. If they are larger than this limit they will be stored in
```

File: man/adoc/bpftrace.adoc [hunk 6]
```diff
@@ -3556,6 +3556,16 @@ The possible options are:
...
+pre-allocated memory.
+
+This threshold is necessary because the BPF stack is limited to 512 bytes and large objects make it more likely for the stack to run out of space. Instead,
+bpftrace can store these large objects in read-write global variables. However, this involves pre-allocating memory so it may be less memory efficient.
+
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:3568]: nit: slight re-wording

```suggestion
This exists because the BPF stack is limited to 512 bytes and large objects make it more likely that we'll run out of space. bpftrace can store objects that are larger than the `on_stack_limit` in pre-allocated memory to prevent this stack error. However, storing in pre-allocated memory may be less memory efficient. Lower this default number if you are still seeing a stack memory error or increase it if you're worried about memory consumption.`

```

File: src/ast/irbuilderbpf.cpp [hunk 7]
```diff
@@ -475,24 +475,61 @@ CallInst *IRBuilderBPF::CreateGetStackScratchMap(StackType stack_type,
                              failure_callback);
 }
 
-Value *IRBuilderBPF::CreateGetStrScratchBuffer(const location &loc, int key)
-{
-  return createScratchBuffer(bpftrace::globalvars::GlobalVar::GET_STR_BUFFER,
-                             loc,
-                             key);
+Value *IRBuilderBPF::CreateGetStrStackOrScratchBufferAllocation(
```
Comment (jordalgo) [inline src/ast/irbuilderbpf.cpp:478]: oofa these function names - what is this Java?. Is something like `CreateGetStrAllocation` not accurate or descriptive enough?

File: src/ast/irbuilderbpf.cpp [hunk 8]
```diff
@@ -475,24 +475,61 @@ CallInst *IRBuilderBPF::CreateGetStackScratchMap(StackType stack_type,
...
+      [&async_ids] { return async_ids.tuple(); });
 }
 
-Value *IRBuilderBPF::CreateTupleScratchBuffer(const location &loc, int key)
+Value *IRBuilderBPF::createStackOrScratchBufferAllocation(
+    bpftrace::globalvars::GlobalVar globalvar,
+    llvm::Type *obj_type,
+    const std::string &name,
+    const location &loc,
+    std::optional<std::function<size_t()>> key_cb)
```
Comment (jordalgo) [inline src/ast/irbuilderbpf.cpp:524]: naming nit
```suggestion
    std::optional<std::function<size_t()>> gen_async_id_cb)
```

**Review by jordalgo — COMMENTED** [review]
Code looks good, just a few small doc updates and naming nits/questions.

**Comment by jordalgo** [comment]
> Changed name to off_stack_limit and updated relevant descriptions

Also, you mean `on_stack_limit`

File: src/ast/irbuilderbpf.cpp (hunk 7 above)
Comment (ryantimwilson) [reply src/ast/irbuilderbpf.cpp:478]: lol yes these are too wordy...i debated changing it and I should've. Yes lets go with Allocation

**Review by jordalgo — APPROVED** [review]
LGTM. Nice work!

**Review by danobi — APPROVED** [review]

File: src/main.cpp [hunk 9]
```diff
@@ -136,6 +136,7 @@ void usage()
   std::cerr << "    BPFTRACE_MAX_PROBES               [default: 512] max number of probes" << std::endl;
   std::cerr << "    BPFTRACE_MAX_STRLEN               [default: 64] bytes on BPF stack per str()" << std::endl;
   std::cerr << "    BPFTRACE_MAX_TYPE_RES_ITERATIONS  [default: 0] number of levels of nested field accesses for tracepoint args" << std::endl;
+  std::cerr << "    BPFTRACE_ON_STACK_LIMIT           [default: 32] max bytes of individual objects to store them on the BPF stack vs. pre-allocated memory" << std::endl;
```
Comment (ajor) [inline src/main.cpp:139]: I don't think this needs to be an environment variable as well as a config option. It seems unlikely that someone would want to change this setting globally for all scripts on their system - it's a thing that'd be very specific to a single script. (this is true for a lot of our env vars, but they exist as historical baggage since they were around before config options were a thing)

File: tests/runtime/call [hunk 10]
```diff
@@ -109,6 +109,24 @@ AFTER ./testprogs/syscall execve /$(python3 -c "print('X'*5555)")
...
+NAME str_scratch_buf
+PROG t:syscalls:sys_enter_execve { print(str(args.filename)) }
+ENV BPFTRACE_ON_STACK_LIMIT=0
+EXPECT /XXXX
+AFTER ./testprogs/syscall execve /$(python3 -c "print('XXXX')")
+REQUIRES_FEATURE probe_read_kernel
+TIMEOUT 1
+
+NAME fmt_str_scratch_buf
+PROG BEGIN { printf("%s %d", "XXXX", 1) }
```
Comment (ajor) [inline tests/runtime/call:126]:
```suggestion
PROG BEGIN { printf("%s %d\n", "XXXX", 1) }
```
Nit: newline to avoid buffering issues (it caught me out when running this script locally at the least)

File: tests/runtime/call [hunk 11]
```diff
@@ -109,6 +109,24 @@ AFTER ./testprogs/syscall execve /$(python3 -c "print('X'*5555)")
...
+NAME str_tuple_scratch_buf
+PROG BEGIN { print((1, (2, "XXXX"))) }
+ENV BPFTRACE_ON_STACK_LIMIT=0
+EXPECT (1, (2, XXXX))
+
+NAME str_scratch_buf
+PROG t:syscalls:sys_enter_execve { print(str(args.filename)) }
+ENV BPFTRACE_ON_STACK_LIMIT=0
+EXPECT /XXXX
+AFTER ./testprogs/syscall execve /$(python3 -c "print('XXXX')")
```
Comment (ajor) [inline tests/runtime/call:121]:
```suggestion
AFTER ./testprogs/syscall execve /XXXX
```
Is the Python part necessary?

File: tests/runtime/call [hunk 12]
```diff
@@ -109,6 +109,24 @@ AFTER ./testprogs/syscall execve /$(python3 -c "print('X'*5555)")
...
+REQUIRES_FEATURE probe_read_kernel
+TIMEOUT 1
+
+NAME fmt_str_scratch_buf
```
Comment (ajor) [inline tests/runtime/call:125]: Sorry, I've probably picked the wrong PR to bring this up on, but the name "fmt_str_scratch_buf" is confusing me. From playing around with it a little, it seems like it's not the format string that's being put into a scratch buf, but rather all the other arguments to `printf`. The format string never appears in BPF code IIUC.

In the case where we're keeping data on the stack, the name "printf_args" is used in place of "fmtstr", which makes more sense to me.

File: tests/runtime/call (hunk 11 above)
Comment (ryantimwilson) [reply tests/runtime/call:121]: No I will fix

File: tests/runtime/call (hunk 10 above)
Comment (ryantimwilson) [reply tests/runtime/call:126]: hmm interesting, I didn't hit it but will fix, thanks!

File: src/main.cpp (hunk 9 above)
Comment (ryantimwilson) [reply src/main.cpp:139]: ack will remove env var

File: tests/runtime/call (hunk 12 above)
Comment (ryantimwilson) [reply tests/runtime/call:125]: technically its called fmt_str_args_scratch_buf, I just fudged the naming here: https://github.com/bpftrace/bpftrace/blob/57b355c95b5406742edeb4b99330e2253703615b/src/required_resources.h#L104

I'll fix the naming here.

I kinda like format string args to printf args but if we want to change it, I prefer we do in another PR

**Comment by ryantimwilson** [comment]
@ajor addressed your comments, mind taking another look?
